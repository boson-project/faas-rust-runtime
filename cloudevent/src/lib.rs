extern crate chrono;
extern crate hostname;
extern crate serde;
extern crate serde_json;
extern crate uuid;

#[macro_use]
extern crate derive_builder;

pub mod http;

use chrono::{DateTime, FixedOffset};
use hostname::get_hostname;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::convert::TryFrom;
use std::fmt;
use uuid::Uuid;

const DEFAULT_TYPE: &str = "generated.rust-faas-runtime";
const DEFAULT_SOURCE: &str = "rustfaasruntime.com";

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub enum SpecVersion {
    #[serde(rename = "0.2")]
    V02,
    #[serde(rename = "0.3")]
    V03,
    #[serde(rename = "1.0")]
    V10,
}

impl fmt::Display for SpecVersion {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SpecVersion::V02 => write!(f, "0.2"),
            SpecVersion::V03 => write!(f, "0.3"),
            SpecVersion::V10 => write!(f, "1.0"),
        }
    }
}

impl TryFrom<String> for SpecVersion {
    type Error = String;

    fn try_from(value: String) -> Result<Self, String> {
        match value.as_str() {
            "0.2" => Ok(SpecVersion::V02),
            "0.3" => Ok(SpecVersion::V03),
            "1.0-rc1" => Ok(SpecVersion::V10),
            _ => Err(format!("Invalid specversion {}", value)),
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub struct Payload {
    #[serde(rename = "datacontenttype")]
    pub content_type: String,

    pub data: String,
}

type PayloadResult<T, E> = Option<Result<T, E>>;

pub trait Writer<T: Sized, E: std::error::Error>
where
    Self: Sized + Clone,
{
    fn write_payload(&mut self, content_type: &str, value: T) -> Result<(), E>;

    fn clone_with_new_payload(&self, content_type: &str, value: T) -> Result<Self, E> {
        let mut new = self.clone();
        new.write_payload(content_type, value)?;
        Ok(new)
    }
}

pub trait Reader<T: Sized, E: std::error::Error> {
    fn read_payload(&self) -> PayloadResult<T, E> {
        self.read_payload_with_content_type()
            .map(|res| res.map(|(_, val)| val))
    }

    fn read_payload_with_content_type(&self) -> PayloadResult<(String, T), E>;
}

pub trait Mapper<T: Sized, E: std::error::Error, F: Fn(T) -> T>
where
    Self: Sized,
{
    fn map_payload(&self, f: F) -> Result<Self, E>;
}

impl<
        T: Sized,
        E: std::error::Error,
        F: Fn(T) -> T,
        S: Writer<T, E> + Reader<T, E> + Clone + Sized,
    > Mapper<T, E, F> for S
{
    fn map_payload(&self, f: F) -> Result<Self, E> {
        if let Some(Ok((ct, value))) = self.read_payload_with_content_type() {
            let mut new = self.clone();
            new.write_payload(&ct, f(value))?;
            return Ok(new);
        } else {
            return Ok(self.clone());
        }
    }
}

#[derive(Serialize, Deserialize, PartialEq, Debug, Clone, Builder)]
#[builder(setter(into, strip_option))]
pub struct Event {
    pub id: String,

    pub source: String,

    #[serde(rename = "specversion")]
    pub spec_version: SpecVersion,

    #[serde(rename = "type")]
    pub event_type: String,

    #[builder(default)]
    pub subject: Option<String>,
    #[builder(default)]
    pub time: Option<DateTime<FixedOffset>>,

    #[serde(flatten)]
    #[builder(default)]
    pub payload: Option<Payload>,

    #[serde(flatten)]
    #[builder(default)]
    pub extensions: HashMap<String, String>,
}

#[allow(non_snake_case)]
impl Event {
    pub fn new() -> Event {
        Event {
            id: Uuid::new_v4().to_string(),
            source: get_hostname().unwrap_or(DEFAULT_SOURCE.to_string()),
            spec_version: SpecVersion::V03,
            event_type: DEFAULT_TYPE.to_string(),
            subject: None,
            time: None,
            payload: None,
            extensions: HashMap::new(),
        }
    }
}

impl Writer<serde_json::Value, serde_json::Error> for Event {
    fn write_payload(
        &mut self,
        content_type: &str,
        value: serde_json::Value,
    ) -> Result<(), serde_json::Error> {
        let serialized = serde_json::to_string(&value)?;
        self.payload = Some(Payload {
            content_type: String::from(content_type),
            data: serialized,
        });
        Ok(())
    }
}

impl Reader<serde_json::Value, serde_json::Error> for Event {
    fn read_payload_with_content_type(
        &self,
    ) -> PayloadResult<(String, serde_json::Value), serde_json::Error> {
        if self.payload.is_none() {
            return None;
        }

        let p = self.payload.as_ref().unwrap();
        Some(
            serde_json::from_str::<serde_json::Value>(p.data.as_str())
                .map(|j| (p.content_type.clone(), j)),
        )
    }
}

impl Reader<serde_json::Value, serde_json::Error> for Option<Event> {
    fn read_payload_with_content_type(
        &self,
    ) -> PayloadResult<(String, serde_json::Value), serde_json::Error> {
        if let Some(r) = self {
            r.read_payload_with_content_type()
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    // Note this useful idiom: importing names from outer (for mod tests) scope.
    use super::*;
    use serde_json::json;

    #[test]
    fn test_serialize_no_payload_no_extensions() {
        let expected_id = "A234-1234-1234";
        let expected_spec_version = SpecVersion::V10;
        let expected_type = "com.github.pull.create";
        let expected_source = "https://github.com/cloudevents/spec/pull";
        let expected_subject = "123";
        let expected_time = DateTime::parse_from_rfc3339("2018-04-05T17:31:00Z").unwrap();

        let j = json!({
            "id" : expected_id,
            "specversion" : expected_spec_version.to_string(),
            "type" : expected_type,
            "source" : expected_source,
            "subject" : expected_subject,
            "time" : expected_time.to_rfc3339()
        });

        let v: Event = serde_json::from_value(j).unwrap();

        assert_eq!(v.id, expected_id);
        assert_eq!(v.spec_version, expected_spec_version);
        assert_eq!(v.event_type, expected_type);
        assert_eq!(v.source, expected_source);
        assert_eq!(v.subject, Some(expected_subject.to_string()));
        assert_eq!(v.time, Some(expected_time));
        assert_eq!(v.payload, None);
        assert!(v.extensions.is_empty());
    }

    #[test]
    fn test_serialize_no_payload_with_extensions() {
        let expected_id = "A234-1234-1234";
        let expected_spec_version = SpecVersion::V10;
        let expected_type = "com.github.pull.create";
        let expected_source = "https://github.com/cloudevents/spec/pull";
        let expected_subject = "123";
        let expected_time = DateTime::parse_from_rfc3339("2018-04-05T17:31:00Z").unwrap();
        let expected_stuff = "aaa";

        let j = json!({
            "id" : expected_id,
            "specversion" : expected_spec_version.to_string(),
            "type" : expected_type,
            "source" : expected_source,
            "subject" : expected_subject,
            "time" : expected_time.to_rfc3339(),
            "stuff": expected_stuff
        });

        let v: Event = serde_json::from_value(j).unwrap();

        assert_eq!(v.id, expected_id);
        assert_eq!(v.spec_version, expected_spec_version);
        assert_eq!(v.event_type, expected_type);
        assert_eq!(v.source, expected_source);
        assert_eq!(v.subject, Some(expected_subject.to_string()));
        assert_eq!(v.time, Some(expected_time));
        assert_eq!(v.payload, None);
        assert!(!v.extensions.is_empty());
        assert_eq!(v.extensions.get("stuff"), Some(&expected_stuff.to_string()));
    }

    #[test]
    fn test_serialize_with_payload_with_extensions() {
        let expected_id = "A234-1234-1234";
        let expected_spec_version = SpecVersion::V10;
        let expected_type = "com.github.pull.create";
        let expected_source = "https://github.com/cloudevents/spec/pull";
        let expected_subject = "123";
        let expected_time = DateTime::parse_from_rfc3339("2018-04-05T17:31:00Z").unwrap();
        let expected_stuff = "aaa";
        let expected_content_type = "application/json";
        let expected_data = r#"{"hello":"world"}"#;

        let j = json!({
            "id" : expected_id,
            "specversion" : expected_spec_version.to_string(),
            "type" : expected_type,
            "source" : expected_source,
            "subject" : expected_subject,
            "time" : expected_time.to_rfc3339(),
            "stuff": expected_stuff,
            "datacontenttype": expected_content_type,
            "data": expected_data
        });

        let v: Event = serde_json::from_value(j).unwrap();

        assert_eq!(v.id, expected_id);
        assert_eq!(v.spec_version, expected_spec_version);
        assert_eq!(v.event_type, expected_type);
        assert_eq!(v.source, expected_source);
        assert_eq!(v.subject, Some(expected_subject.to_string()));
        assert_eq!(v.time, Some(expected_time));
        assert_eq!(
            v.payload,
            Some(Payload {
                content_type: expected_content_type.to_string(),
                data: expected_data.to_string()
            })
        );
        assert!(!v.extensions.is_empty());
        assert_eq!(v.extensions.get("stuff"), Some(&expected_stuff.to_string()));
    }
}
