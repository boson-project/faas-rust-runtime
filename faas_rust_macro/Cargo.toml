[package]
name = "faas_rust_macro"
version = "0.1.0"
authors = ["Francesco Guardiani <francescoguard@gmail.com>"]
edition = "2018"

[lib]
proc-macro = true

[dependencies]
syn = { version = "^1", features = ["full", "extra-traits"] }
quote = "^1"
proc-macro2 = "^1"
regex = "1.3.3"

[dev-dependencies]
trybuild = "^1.0"

# Needed to compile the test functions
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
futures = "^0.3"
actix-web = "2.0.0"
actix-rt = "1.0.0"
faas_rust = { path = "../faas_rust"}
cloudevent = { path = "../cloudevent" }
