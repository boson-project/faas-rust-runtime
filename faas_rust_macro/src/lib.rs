extern crate proc_macro;

use proc_macro2::TokenStream;
use quote::{format_ident, quote, quote_spanned};
use std::borrow::Borrow;
use syn::{spanned::Spanned, FnArg, Ident, ReturnType, Type};

mod types;

#[proc_macro_attribute]
pub fn faas_function(
    _args: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let function_ast: syn::ItemFn = syn::parse(item.clone()).unwrap();
    let out = impl_faas_function(function_ast);
    out.into()
}

fn impl_faas_function(user_function: syn::ItemFn) -> TokenStream {
    let main_fn: TokenStream = quote! {
        #[actix_rt::main]
        async fn main() -> std::io::Result<()> {
            faas_rust::start_runtime(|r| r.to(handle_event)).await
        }
    };
    let handler = generate_handler(user_function.clone());

    quote! {
        use std::iter::FromIterator;

        #user_function
        #handler
        #main_fn
    }
}

fn generate_handler(function_ast: syn::ItemFn) -> TokenStream {
    let user_function_name = function_ast.sig.ident.clone();

    // Function input

    let input_extracted: Vec<(Ident, TokenStream)> = function_ast.sig.inputs
        .iter()
        .enumerate()
        .map(|(i, arg)|
            extract_type_from_fn_arg(arg)
                .and_then(|ty| {
                    let varname = format_ident!("_arg{}", i);
                    if is_event(ty) {
                        let num = i + 1;
                        Some((varname.clone(), quote_spanned! {arg.span()=>
                            let #varname: cloudevent::Event = events.pop().ok_or(actix_web::error::ErrorBadRequest(format!("Expecting event in position {}", #num)))?;
                        }))
                    } else if is_option_event(ty) {
                        Some((varname.clone(), quote_spanned! {arg.span()=>
                            let #varname: Option<cloudevent::Event> = events.pop();
                        }))
                    } else {
                        None
                    }
                })
                .unwrap_or((
                    format_ident!("{}", "err"),
                    syn::Error::new_spanned(arg, "Type should be Event or Option<Event>").to_compile_error()
                ))

        )
        .collect();

    let (input_extracted_ident, input_extracted_stmts): (Vec<Ident>, Vec<TokenStream>) =
        input_extracted.iter().cloned().unzip();

    // Function invocation

    let mut user_function_invocation = quote! {
            #user_function_name(#(#input_extracted_ident),*)
    };

    if function_ast.sig.asyncness.is_some() {
        user_function_invocation = quote! {
            #user_function_invocation.await
        }
    };

    // Function output

    let output_mapper: TokenStream = map_output(&function_ast.sig.output).unwrap_or(
        syn::Error::new_spanned(
            function_ast.sig,
            "Return type should be Result<V, E>, where V is Vec<Event> or Option<Event> or Event",
        )
        .to_compile_error(),
    );

    // fn handleEvent()

    let out = quote! {
            async fn handle_event(
            req: actix_web::HttpRequest,
            body: actix_web::web::Bytes,
        ) -> Result<actix_web::HttpResponse, actix_web::Error> {
            let value = faas_rust::request_reader::read_cloud_event(req, body).await?;

            // Unzip
            let (encoding, mut events) = match value {
                Some((encoding, events)) => (Some(encoding), events),
                None => (None, vec![])
            };

            events.reverse();

            #(#input_extracted_stmts)*

            let output = #user_function_invocation?;
            let mapped_output: Vec<cloudevent::Event> = #output_mapper;
            faas_rust::response_writer::write_cloud_event(mapped_output, encoding)
        }
    };

    out.into()
}

fn map_output(rt: &ReturnType) -> Option<TokenStream> {
    let result_type = match rt {
        ReturnType::Type(_, ty) => types::extract_types_from_result(&ty),
        _ => None,
    }?;
    let result_left = result_type.0;

    if is_vec_event(result_left) {
        Some(quote! {
        output
        })
    } else if is_option_event(result_left) {
        Some(quote! {
        Vec::from_iter(output.into_iter());
        })
    } else if is_event(result_left) {
        Some(quote! {
        vec![output]
        })
    } else {
        None
    }
}

fn is_vec_event(ty: &Type) -> bool {
    let extracted = types::extract_types_from_vec(ty);
    let type_matcher = types::generate_type_matcher("cloudevent::Event");
    match extracted {
        Some(t) => type_matcher(t),
        _ => false,
    }
}

fn is_option_event(ty: &Type) -> bool {
    let extracted = types::extract_types_from_option(ty);
    let type_matcher = types::generate_type_matcher("cloudevent::Event");
    match extracted {
        Some(t) => type_matcher(t),
        _ => false,
    }
}

fn is_event(ty: &Type) -> bool {
    types::generate_type_matcher("cloudevent::Event")(ty)
}

fn extract_type_from_fn_arg(fn_arg: &FnArg) -> Option<&Type> {
    match fn_arg {
        FnArg::Typed(ty) => Some(ty.ty.borrow()),
        _ => None,
    }
}
