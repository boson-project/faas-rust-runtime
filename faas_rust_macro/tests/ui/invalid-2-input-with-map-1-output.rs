#![feature(start)]

use cloudevent::Event;
use faas_rust_macro::faas_function;

#[faas_function]
pub async fn function(
    last: std::collections::HashMap<String, Event>,
    other: Event,
) -> Result<Event, actix_web::Error> {
    Ok(other)
}

#[start]
fn start(_argc: isize, _argv: *const *const u8) -> isize {
    0
}
