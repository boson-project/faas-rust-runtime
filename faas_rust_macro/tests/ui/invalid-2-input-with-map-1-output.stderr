error: Type should be Event or Option<Event>
 --> $DIR/invalid-2-input-with-map-1-output.rs:8:5
  |
8 |     last: std::collections::HashMap<String, Event>,
  |     ^^^^

error[E0308]: mismatched types
 --> $DIR/invalid-2-input-with-map-1-output.rs:6:1
  |
6 | #[faas_function]
  | ^^^^^^^^^^^^^^^^ expected struct `std::collections::HashMap`, found enum `std::option::Option`
  |
  = note: expected struct `std::collections::HashMap<std::string::String, cloudevent::Event>`
               found enum `std::option::Option<cloudevent::Event>`
